invokeai/backend/model_manager/install.py

```diff
@@ -0,0 +1,758 @@
+        if self._tmpdir and len(jobs) <= 1 and job.status in ["completed", "error", "cancelled"]:
+            self._tmpdir.cleanup()
+            self._tmpdir = None
+
+    def _complete_registration_handler(self, job: DownloadJobBase):
+        job = ModelInstallJob.parse_obj(job)  # upcast should succeed
+        if job.status == "completed":
+            self._logger.info(f"{job.source}: Installing in place.")
+            model_id = self.register_path(job.destination, job.probe_override)
+            info = self._store.get_model(model_id)
+            info.source = str(job.source)
+            info.description = f"Imported model {info.name}"
+            self._store.update_model(model_id, info)
+            self._async_installs[job.source] = model_id
+            job.model_key = model_id
+        elif job.status == "error":
+            self._logger.warning(f"{job.source}: Model installation error: {job.error}")
+        elif job.status == "cancelled":
+            self._logger.warning(f"{job.source}: Model installation cancelled at caller's request.")
+
+    def sync_model_path(self, key) -> ModelConfigBase:
+        """
+        Move model into the location indicated by its basetype, type and name.
+
+        Call this after updating a model's attributes in order to move
+        the model's path into the location indicated by its basetype, type and
+        name. Applies only to models whose paths are within the root `models_dir`
+        directory.
+
+        May raise an UnknownModelException.
+        """
+        model = self._store.get_model(key)
+        old_path = Path(model.path)
+        models_dir = self._app_config.models_path
+
+        if not old_path.is_relative_to(models_dir):
+            return model
+
+        new_path = models_dir / model.base_model.value / model.model_type.value / model.name
+        self._logger.info(
+            f"{old_path.name} is not in the right directory for a model of its type. Moving to {new_path}."
+        )
+        model.path = self._move_model(old_path, new_path).as_posix()
+        self._store.update_model(key, model)
+        return model
+
+    def _make_download_job(
+        self,
+        source: Union[str, Path, AnyHttpUrl],
+        variant: Optional[str] = None,
+        access_token: Optional[str] = None,
+        priority: Optional[int] = 10,
+    ) -> DownloadJobBase:
+        # Clean up a common source of error. Doesn't work with Paths.
+        if isinstance(source, str):
+            source = source.strip()
+
+        # In the event that we are being asked to install a path that is already on disk,
+        # we simply probe and register/install it. The job does not actually do anything, but we
+        # create one anyway in order to have similar behavior for local files, URLs and repo_ids.
+        if Path(source).exists():  # a path that is already on disk
+            destdir = source
+            return ModelInstallPathJob(source=source, destination=Path(destdir))
```

Comment:
Does this re-hash the model to ensure integrity?

Reply:
When the user asks to register a model that is already on local disk it gets probed and the hash is generated. Then the path and hash are entered into the config record database without changing the location of the model. So rehashing not needed.

Alternatively, the user can elect to install the model, in which case the code moves the model from wherever it currently is into the appropriate location in the `models` directory. The hash is calculated before the model is moved into `models`, but not after. I would expect the operating system to give an error if the copy operation fails in some way, but I'll add the extra hashing just to be safe.

BTW, you helped me find a bug here. The code uses `rename()` to move the model, which won't work across filesystems.